### PR TITLE
On Windows and macOS, add API to enable/disable window controls

### PR DESCRIPTION
**Pull request by amrbashir**
- [X] Tested on all platforms changed
	- [X] Windows
	- [x] macOS
- [X] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [X] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [X] Created or updated an example program if it would help users understand this functionality
- [X] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Porting https://github.com/tauri-apps/tao/commit/a50fd867b3df033dff077f5320b4b7037b04e454

### CODE REVIEW

**Comment by kchibisov** [comment]
I have a general question wrt that. Is this about the decoration thing or you can actually disable `Minimize`/`Maximize`, ect? What the goal for `set_closable`, etc? If that's all about buttons and has nothing to do with the ability to close the window it should be named differently. From now I see that `set_closable(false)` prevents the window close?

**Comment by amrbashir** [comment]
Yeah, this is mainly about disabling the buttons itself so it can't be clicked but the window can still be closed in other ways.

**Comment by kchibisov** [comment]
Then we should explicitly say that it's all about buttons in decorations. You can use `bitflags` setter and getter here, no need for so many methods returning bool. It's also better to manage bitflags than a bunch of bools.

**Review by kchibisov — CHANGES_REQUESTED** [review]
See above.

The bitflags should be used instead.
The api should explicitly mention decorations.

Sorry for that short of review here, don't have that many of time and I'm only capable to reveiw general changes in the API not the windows/macOS code anyway.

**Comment by amrbashir** [comment]
Done, although I don't really like the naming, so I am open to suggestions.

File: src/window.rs [hunk 1]
```diff
@@ -732,6 +745,26 @@ impl Window {
         self.window.is_resizable()
     }
 
+    /// Sets the enabled window buttons.
+    ///
+    /// ## Platform-specific
+    ///
+    /// - **Wayland / X11:** Not implemented.
+    /// - **Web / iOS / Android:** Unsupported.
+    pub fn set_window_buttons(&self, buttons: WindowButtons) {
```
Comment (madsmtm) [inline src/window.rs:754]: Naming idea: `[set_]enabled_buttons` or `[set_]visible_buttons` (`window` is at least redundant here)

**Review by madsmtm — COMMENTED** [review]
Thanks! I agree with the bitflag API.

An example would be nice for testing the functionality?

File: src/window.rs (hunk 1 above)
Comment (amrbashir) [reply src/window.rs:754]: `[set_]enabled_buttons` sounds good, thanks.

**Review by kchibisov — APPROVED** [review]
Api change is fine. The rest is up to windows/macOS maintainers.

**Review by madsmtm — CHANGES_REQUESTED** [review]
macOS impl looks good, haven't tested it though; would you mind adding an example for it?

**Comment by amrbashir** [comment]
Sure

**Review by madsmtm — APPROVED** [review]
Thanks for the example - it enabled me to find a bug which I've pushed a fix for ;)

File: src/platform_impl/windows/window.rs [hunk 2]
```diff
@@ -974,6 +1016,9 @@ where
     window_flags.set(WindowFlags::TRANSPARENT, attributes.transparent);
     // WindowFlags::VISIBLE and MAXIMIZED are set down below after the window has been configured.
     window_flags.set(WindowFlags::RESIZABLE, attributes.resizable);
+    // Will be changed later using `window.set_enabled_buttons` but we need to set a default here
+    // so the diffing later can work
```
Comment (kchibisov) [inline src/platform_impl/windows/window.rs:1020]:
```suggestion
    // Will be changed later using `window.set_enabled_buttons` but we need to set a default here
    // so the diffing later can work.
```

File: src/platform_impl/windows/window_state.rs [hunk 3]
```diff
@@ -344,6 +349,21 @@ impl WindowFlags {
...
+                EnableMenuItem(
+                    GetSystemMenu(window, 0),
+                    SC_CLOSE,
+                    MF_BYCOMMAND
+                        | if new.contains(WindowFlags::CLOSABLE) {
+                            MF_ENABLED
+                        } else {
+                            MF_DISABLED
+                        },
+                );
```
Comment (kchibisov) [inline src/platform_impl/windows/window_state.rs:363]: Could we compose this on a separate line, so this is a bit more readable?

**Review by kchibisov — CHANGES_REQUESTED** [review]
Just a few minor notes. Should be fine afterwards.

**Review by msiglreith — APPROVED** [review]
Tested successfully on Windows 10
(interesting that Windows seems to automatically adjust the style towards a tool window once minimize and maximize are disabled)

File: src/platform_impl/windows/window_state.rs (hunk 3 above)
Comment (amrbashir) [reply src/platform_impl/windows/window_state.rs]: separating it on a new line didn't feel to me as much enhancement so I too the liberty to use `bool::then` in [`4ff0733`](https://github.com/rust-windowing/winit/pull/2537/commits/4ff07333d7a68fdf655996e5962dcc7c8c1c579f)

**Review by kchibisov — APPROVED** [review]
Could you rebase? Should be good to go afterwards.

**Comment by amrbashir** [comment]
Done!

**Pull request merged**